### PR TITLE
Oop exercise

### PR DESCRIPTION
**Pull request by igkostyuk**
### OOP exercise

[Demo](https://igkostyuk.github.io/a-tiny-JS-world/) |
[Code base](https://github.com/igkostyuk/a-tiny-JS-world)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

**Comment by OleksiyRudenko** [comment]
@igkostyuk Why did you merge `master` into the feature branch?

**Comment by OleksiyRudenko** [comment]
@igkostyuk At a glance you did pretty well.
Please, check yourself against the following list:
Minimal requirements to meet:
 * [ ] Implement a base class to inherit from
 * [ ] Employ default parameters
 * [ ] Each species is represented with its own class
 * [ ] No need to specify species at instantiation
 * [ ] Classes for species that do not have hands by natural design
       do not consequently have `hands` or any equivalent property
 * [ ] All inhabitants are stored in a container (array or object)
 * [ ] JS native features are intensively employed (`const`, `let`, `Array.map|join|forEach|...`, etc)
 * [ ] [OOP, SOLID and DRY](https://github.com/OleksiyRudenko/a-tiny-JS-world/blob/master/README.md#learn-on-your-own)
       principles are intensively employed

Please, confirm all items above are positive. 
 
Optional level up (not required to implement):
 * [ ] Friends list is a list of objects refs rather than names (strings)
 * [ ] Cat-woman class is built employing composition rather than inheritance only

[ES6 classes cheat-sheet](https://gist.github.com/OleksiyRudenko/672d39b08d9d0da4e179aca49876c58b)

Please, ask a question if anything above requires clarification.

**Comment by igkostyuk** [comment]
tried to solve https://github.com/kottans/frontend-2019-homeworks/pull/121
already checked

File: submissions/igkos/OOP-exercise/index.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    }; ${this.saying}; ${this.friends}`;
+  }
+}
+
+class Human extends Inhabitant {
+  constructor(name, saying, friends, species, legs, hands, gender) {
+    super(name, saying, friends, species, legs, hands, gender);
+    this.species = "human";
+    this.legs = 2;
+    this.hands = 2;
```
Comment (OleksiyRudenko) [inline submissions/igkos/OOP-exercise/index.js]: As you override legs and hands count anyway. Why (1) asking user to provide values at construction time and (2) not supplying these values to `super`?
Same questions re `species`.
And other properties across the code.
Please, fix.

File: submissions/igkos/OOP-exercise/index.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    this.gender = gender;
+    this.legs = legs;
+    this.hands = hands;
+    this.saying = saying;
+    this.friends = friends;
+  }
+  toString() {
+    return `${this.species}; ${this.name}; ${this.gender}; ${this.legs}; ${
+      this.hands
+    }; ${this.saying}; ${this.friends}`;
```
Comment (OleksiyRudenko) [inline submissions/igkos/OOP-exercise/index.js]: Do you really need `this.hands` on a separate row?
Your construction can be optimized using `#Array.join()`
Please, fix.

File: submissions/igkos/OOP-exercise/index.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    super(name, saying, friends, species, legs, hands, gender);
+    this.species = "cat";
+    this.saying = "meow-meow!";
+  }
+}
+
+[
+  new Female("Jane", "Hey! I am a STRONG WOMAN", ["Artur", "Tomy"]),
+  new Male("Artur", "All right everyone, listen up.", ["Jane", "Toby"]),
+  new Cat("Tomy", "female", ["Jane"]),
```
Comment (OleksiyRudenko) [inline submissions/igkos/OOP-exercise/index.js]: If you specify friends these must be other existing objects, not just strings denoting names.
Please, fix.

File: submissions/igkos/OOP-exercise/index.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  new Dog("Toby", "male", ["Artur"])
+].forEach(inhabitant => print(inhabitant.toString()));
```
Comment (OleksiyRudenko) [inline submissions/igkos/OOP-exercise/index.js]: Try what happens if you do not specify `.toString` explicitly. That's the point of this magic method.
Please, read more about `toString`.

File: submissions/igkos/OOP-exercise/index.js (hunk 2 above)
Comment (igkostyuk) [reply submissions/igkos/OOP-exercise/index.js]: Prettie move this.hands

File: submissions/igkos/OOP-exercise/index.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+class Animal extends Inhabitant {
+  constructor(name, saying, friends, species, legs  = 4, hands = 0, gender) {
+    super(name, saying, friends, species, legs, hands, gender);
+  }
+}
+
+class Dog extends Animal {
+  constructor(name, gender, friends, species = "Dog", legs, hands, saying = "woof-woof!") {
```
Comment (OleksiyRudenko) [inline submissions/igkos/OOP-exercise/index.js]: When a user creates a dog (`new Dog(...)`) you don't need to ask for a species as a parameter to pass on.
When calling `super` you can explicitly specify the species.
Please, fix for `Dog` and for other species as well.

**Comment by igkostyuk** [comment]
Please, review

File: submissions/igkos/OOP-exercise/index.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+const Inhabitants = [
+  new Female("Jane", "Hey! I am a STRONG WOMAN"),
+  new Male("Artur", "All right everyone, listen up."),
+  new Cat("Tomy", "female"),
+  new Dog("Toby", "male")
+]
+
+Inhabitants.forEach(Inhabitant => {
+  Inhabitant.setFriends(Inhabitants.filter(currentName => currentName.name !== Inhabitant.name).map(friend => friend.name));
+  print(Inhabitant);
```
Comment (OleksiyRudenko) [inline submissions/igkos/OOP-exercise/index.js]: You do not need to fix this, just keep in mind.

There is a Separation of Concerns principle, which says that a single unit of code has to deal with a single major problem.

If applied to this fragment of code it means that adding friends (which is a part of objects initialization) should be separated from output (which is a part of interaction with an end-user).

Following the separation of concerns principle helps building scalable code.

**Pull request merged**